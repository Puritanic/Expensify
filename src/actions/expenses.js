import database from '../firebase/firebase';

// * Actions workflow
// ? component calls action generator
// ? action generator returns function
// ? component dispatches function (redux middleware)(redux-thunk?)
// ? redux store changes

// ADD_EXPENSE
export const addExpense = expense => ({
    type: 'ADD_EXPENSE',
    expense,
});

export const startAddExpense = (expenseData = {}) => (dispatch) => {
    // set defaults and destructure from expenseData
    const {
        description = '',
        notes = '',
        amount = 0,
        createdAt = 0,
    } = expenseData;
    const expense = {
        description,
        notes,
        amount,
        createdAt,
    };
    return database
        .ref('expenses')
        .push(expense)
        .then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense,
            }), );
        });
};
// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id,
});
// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates,
});
