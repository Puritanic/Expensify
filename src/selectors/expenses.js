import moment from 'moment';

const getVisibleExpenses = (expenses, {
    text, sortBy, startDate, endDate
}) =>
    expenses
        .filter((expense) => {
            const createdAtMoment = moment(expense.createdAt);
            const startDateMatch = startDate
                ? startDate.isSameOrBefore(createdAtMoment, 'day')
                : true;
            const endDateMatch = endDate
                ? endDate.isSameOrAfter(createdAtMoment, 'day')
                : true;
            const textMatch = expense.description
                .toLowerCase()
                .includes(text.toLowerCase());

            return startDateMatch && endDateMatch && textMatch;
        })
        // eslint-disable-next-line array-callback-return
        .sort((a, b) => {
            // from highest to lowest value
            if (sortBy === 'date') {
                return a.createdAt < b.createdAt ? 1 : -1;
            } else if (sortBy === 'amount') {
                return a.amount < b.amount ? 1 : -1;
            }
        });

export default getVisibleExpenses;
