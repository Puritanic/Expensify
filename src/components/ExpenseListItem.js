import React from "react";
import { connect } from "react-redux";
import { removeExpense } from "../actions/expenses";

const ExpenseListItem = ({ description, amount, createdAt, id, dispatch }) => (
    <div>
        <h4>{description}</h4>
        <p>
            {amount} -- {new Date(createdAt * 1000).toDateString()}
            <button
                onClick={() => {
                    dispatch(removeExpense({ id }));
                }}
            >
                &times;
            </button>
        </p>
    </div>
);
/* NOTE: In the first pair of the parenthesis we re getting state info,
    but for this operation we dont need to know anything about state, so we can leave it blank.
*/

export default connect()(ExpenseListItem);
