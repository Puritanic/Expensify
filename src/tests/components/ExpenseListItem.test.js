import React from 'react';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';
/* eslint-disable no-undef */

it('should render ExpenseListItem', () => {
    const wrapper = shallow(<ExpenseListItem key={expenses[1].id} {...expenses[1]} />, );
    expect(toJson(wrapper)).toMatchSnapshot();
});
