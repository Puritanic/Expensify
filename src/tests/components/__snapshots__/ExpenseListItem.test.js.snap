// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/2"
  >
    <h4>
      Cable
    </h4>
  </Link>
  <p>
    $10.00
     -- 
    Dec 28th, 1969
  </p>
</div>
`;
